cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,329 @@
+		pluginName, err := notationplugin.ExtractPluginNameFromFileName(f.Name)
+		if err != nil {
+			logger.Infof("processing file %s. File name not in format notation-{plugin-name}, skipped", f.Name)
+			continue
+		}
+		fileInArchive, err := f.Open()
+		if err != nil {
+			return err
+		}
+		defer fileInArchive.Close()
+		return installPluginExecutable(ctx, f.Name, pluginName, fileInArchive, force)
+	}
+	return fmt.Errorf("no valid plugin file was found in %s. Plugin file name must in format notation-{plugin-name}", zipPath)
+}
+
+// installPluginFromTarGz extracts and untar a plugin tar.gz file, validates and
+// installs the plugin
+func installPluginFromTarGz(ctx context.Context, tarGzPath string, force bool) error {
+	logger := log.GetLogger(ctx)
+	rc, err := os.Open(tarGzPath)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	decompressedStream, err := gzip.NewReader(rc)
+	if err != nil {
+		return err
+	}
+	defer decompressedStream.Close()
+	tarReader := tar.NewReader(decompressedStream)
+	// require one and only one file with name in the format
+	// `notation-{plugin-name}`
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if !header.FileInfo().Mode().IsRegular() || strings.Contains(header.Name, "..") {
+			continue
+		}
+		// validate and get plugin name from file name
+		pluginName, err := notationplugin.ExtractPluginNameFromFileName(header.Name)
+		if err != nil {
+			logger.Infof("processing file %s. File name not in format notation-{plugin-name}, skipped", header.Name)
+			continue
+		}
+		return installPluginExecutable(ctx, header.Name, pluginName, tarReader, force)
+	}
+	return fmt.Errorf("no valid plugin file was found in %s. Plugin file name must in format notation-{plugin-name}", tarGzPath)
+}
+
+// installPluginExecutable extracts, validates, and installs a plugin file
+func installPluginExecutable(ctx context.Context, fileName string, pluginName string, fileReader io.Reader, force bool) error {
+	// sanity check
+	if runtime.GOOS == "windows" && filepath.Ext(fileName) != ".exe" {
+		return fmt.Errorf("on Windows, plugin executable file %s is missing the '.exe' extension", fileName)
+	}
+	if runtime.GOOS != "windows" && filepath.Ext(fileName) == ".exe" {
+		return fmt.Errorf("on %s, plugin executable file %s cannot have the '.exe' extension", runtime.GOOS, fileName)
+	}
+	// extract to tmp dir
+	tmpDir, err := os.MkdirTemp(".", notationPluginTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to create notationPluginTmpDir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpFilePath := filepath.Join(tmpDir, fileName)
+	pluginFile, err := os.OpenFile(tmpFilePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0700)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(pluginFile, fileReader); err != nil {
+		defer pluginFile.Close()
+		return err
+	}
+	if err := pluginFile.Close(); err != nil {
+		return err
```

Comment:
The previous close happens when there's an error during `io.Copy`, if we do not close `pluginFile` there, `defer os.RemoveAll(tmpDir)` would fail silently.
The reason we need to manually close `pluginFile` here is that the file is used again in `GetPluginMetadata`, we need to close it before using it again.